client/write.go

```diff
@@ -117,25 +121,54 @@ func (w *writeState) reset() {
 func (w *writeState) close() {
 	w.session.writeOpPool.Put(w.op)
 
-	w.op, w.majority, w.successful, w.pending = nil, 0, 0, 0
+	w.op, w.majority, w.pending, w.halfSuccess = nil, 0, 0, 0
 	w.errors = w.errors[:0]
 	w.queues = w.queues[:0]
 
 	w.session.writeStatePool.Put(w)
 }
 
+func (w *writeState) addError(err error) {
+	w.Lock()
+	w.errors = append(w.errors, err)
+	w.Unlock()
+}
+
+// todo@bl - modify comment on what happens to inflight writes
 func (w *writeState) completionFn(result interface{}, err error) {
 	var (
 		successful int32
 		remaining  = atomic.AddInt32(&w.pending, -1)
+		hostID     = result.(string)
+		// NB(bl) panic on invalid result, it indicates a bug in the code
 	)
 
 	if err != nil {
-		w.Lock()
-		w.errors = append(w.errors, err)
-		w.Unlock()
+		w.addError(err)
+	} else if hostShardSet, ok := w.topoMap.LookupHostShardSet(hostID); !ok {
+		w.addError(fmt.Errorf("Missing host shard in writeState completionFn: %s", hostID))
+	} else if shardState, err :=
+		hostShardSet.ShardSet().LookupStateByID(w.op.shardID); err != nil {
+		w.addError(fmt.Errorf("Missing shard %d in host %s", w.op.shardID, hostID))
 	} else {
-		successful = atomic.AddInt32(&w.successful, 1)
+		// NB(bl): We use one variable, halfSuccess, to avoid atomically
```

Comment:
We should avoid multiple atomics to prevent races. I don't have a strong opinion about sticking with one atomic or adding more locking, though I prefer the former for being less code. 

@other-reviewers Please lmk if the code is too confusing as-is, and I'll update it.